presto-main/src/main/java/com/facebook/presto/memory/QueryContext.java

```diff
@@ -81,6 +85,34 @@ public synchronized void freeMemory(long bytes)
     {
         checkArgument(reserved - bytes >= 0, "tried to free more memory than is reserved");
         reserved -= bytes;
+        memoryPool.free(bytes);
+    }
+
+    public synchronized void setMemoryPool(MemoryPool pool)
+    {
+        requireNonNull(pool, "pool is null");
+        if (pool.getId().equals(memoryPool.getId())) {
+            // Don't unblock our tasks and thrash the pools, if this is a no-op
+            return;
+        }
+        memoryPool.free(reserved);
```

Comment:
you are still freeing before reserving the new memory.  I'd put the `free` in the callback.